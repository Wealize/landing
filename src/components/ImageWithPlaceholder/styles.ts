import styled from 'styled-components'

export const Container = styled.div`
  display: flex;
  position: relative;
  width: 100%;
  height: 100%;
`
