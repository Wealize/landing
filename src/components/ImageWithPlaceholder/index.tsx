import React, { CSSProperties, useState } from 'react'
import Image from 'next/image'
import ReactPlaceholder from 'react-placeholder'

import { Container } from './styles'

type ImageWithPlaceholderProps = {
  imageUrl: string,
  imageAriaLabel?: string,
  imageTitle?: string,
  imageQuality?: number,
  objectFit?: 'cover' | 'contain' | 'fill' | 'none' | 'scale-down'
  showLoadingAnimation?: boolean,
  placeholderType?: 'text' | 'media' | 'textRow' | 'rect' | 'round',
  style?: CSSProperties
}

const ImageWithPlaceholder = (props: ImageWithPlaceholderProps) => {
  const {
    imageUrl,
    imageAriaLabel,
    imageTitle,
    imageQuality = 75,
    objectFit = 'cover',
    showLoadingAnimation = true,
    placeholderType = 'rect',
    style
  } = props
  const [showPlaceholder, setIsShowPlaceholder] = useState(false)
  const DELAY_FOR_HIDE_PLACEHOLDER = 3000
  return (
    <Container style={style}>
      <ReactPlaceholder
        type={placeholderType}
        firstLaunchOnly={true}
        showLoadingAnimation={showLoadingAnimation}
        ready={showPlaceholder}
        style={{ marginRight: '0', position: 'absolute' }}
        // eslint-disable-next-line react/no-children-prop
        children={null}
      />
      <Image
        src={imageUrl}
        layout="fill"
        objectFit={objectFit}
        aria-label={imageAriaLabel}
        quality={imageQuality}
        title={imageTitle}
        onLoad={() => {
          setTimeout(() => {
            setIsShowPlaceholder(true)
          }, DELAY_FOR_HIDE_PLACEHOLDER)
        }}
      />
    </Container>
  )
}

export default React.memo(ImageWithPlaceholder)
