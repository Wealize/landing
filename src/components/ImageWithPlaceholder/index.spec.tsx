import React from 'react'
import { mount } from 'enzyme'
import Image from 'next/image'

import ImageWithPlaceholder from './index'

describe('<ImageWithPlaceholder />', () => {
  const imageWithPlaceholderProps = {
    imageUrl: 'https://commons.wikimedia.org/wiki/File:Old_version_of_the_Wikipedia_logo_used_until_2010_(big,_English).png',
    imageAriaLabel: 'image-aria-label',
    imageTitle: 'image-title',
    imageQuality: 75,
    showLoadingAnimation: true,
    placeholderType: 'round' as const
  }
  const wrapper = mount(<ImageWithPlaceholder {...imageWithPlaceholderProps} />)
  const imageWithPlaceholderWithoutOptionalProps = {
    imageUrl: 'https://commons.wikimedia.org/wiki/File:Old_version_of_the_Wikipedia_logo_used_until_2010_(big,_English).png'
  }
  const wrapperWithoutOptionalProps = mount(<ImageWithPlaceholder {...imageWithPlaceholderWithoutOptionalProps} />)

  it('should render', () => {
    expect(wrapper).toBeDefined()
    expect(wrapperWithoutOptionalProps).toBeDefined()
    expect(wrapper).toMatchSnapshot()
    expect(wrapperWithoutOptionalProps).toMatchSnapshot()
  })

  it('should render content correctly', () => {
    expect(wrapper.find('img').length).toEqual(1)
    expect(wrapper.find('img').prop('aria-label')).toEqual(imageWithPlaceholderProps.imageAriaLabel)
    expect(wrapper.find('img').prop('title')).toEqual(imageWithPlaceholderProps.imageTitle)
    expect(wrapper.find(Image).prop('src')).toEqual(imageWithPlaceholderProps.imageUrl)
    expect(wrapper.find('.round-shape').length).toEqual(1)
    expect(wrapper.find('div.show-loading-animation').length).toEqual(1)
  })


  it('should render content correctly without optional props', () => {
    expect(wrapperWithoutOptionalProps.find('img').length).toEqual(1)
    expect(wrapperWithoutOptionalProps.find('img').prop('aria-label')).toEqual(undefined)
    expect(wrapperWithoutOptionalProps.find('img').prop('title')).toEqual(undefined)
    expect(wrapperWithoutOptionalProps.find(Image).prop('src')).toEqual(imageWithPlaceholderProps.imageUrl)
    expect(wrapperWithoutOptionalProps.find('.rect-shape').length).toEqual(1)
    expect(wrapperWithoutOptionalProps.find('div.show-loading-animation').length).toEqual(1)
  })
})
