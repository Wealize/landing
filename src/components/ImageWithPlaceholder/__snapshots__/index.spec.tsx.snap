// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ImageWithPlaceholder /> should render 1`] = `
<Memo(ImageWithPlaceholder)
  imageAriaLabel="image-aria-label"
  imageQuality={75}
  imageTitle="image-title"
  imageUrl="https://commons.wikimedia.org/wiki/File:Old_version_of_the_Wikipedia_logo_used_until_2010_(big,_English).png"
  placeholderType="round"
  showLoadingAnimation={true}
>
  <styles__Container>
    <div
      className="styles__Container-sc-1li6v9v-0 HMjkw"
    >
      <ReactPlaceholder
        firstLaunchOnly={true}
        ready={false}
        showLoadingAnimation={true}
        style={
          Object {
            "marginRight": "0",
            "position": "absolute",
          }
        }
        type="round"
      >
        <RoundShape
          className="show-loading-animation"
          color="#CDCDCD"
          rows={3}
          style={
            Object {
              "marginRight": "0",
              "position": "absolute",
            }
          }
        >
          <div
            className="round-shape show-loading-animation"
            style={
              Object {
                "backgroundColor": "#CDCDCD",
                "borderRadius": "500rem",
                "height": "100%",
                "marginRight": "0",
                "position": "absolute",
                "width": "100%",
              }
            }
          />
        </RoundShape>
      </ReactPlaceholder>
      <Image
        aria-label="image-aria-label"
        layout="fill"
        objectFit="cover"
        onLoad={[Function]}
        quality={75}
        src="https://commons.wikimedia.org/wiki/File:Old_version_of_the_Wikipedia_logo_used_until_2010_(big,_English).png"
        title="image-title"
      >
        <div
          style={
            Object {
              "bottom": 0,
              "boxSizing": "border-box",
              "display": "block",
              "left": 0,
              "margin": 0,
              "overflow": "hidden",
              "position": "absolute",
              "right": 0,
              "top": 0,
            }
          }
        >
          <img
            aria-label="image-aria-label"
            decoding="async"
            onLoad={[Function]}
            src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
            style={
              Object {
                "border": "none",
                "bottom": 0,
                "boxSizing": "border-box",
                "display": "block",
                "height": 0,
                "left": 0,
                "margin": "auto",
                "maxHeight": "100%",
                "maxWidth": "100%",
                "minHeight": "100%",
                "minWidth": "100%",
                "objectFit": "cover",
                "objectPosition": undefined,
                "padding": 0,
                "position": "absolute",
                "right": 0,
                "top": 0,
                "visibility": "hidden",
                "width": 0,
              }
            }
            title="image-title"
          />
        </div>
      </Image>
    </div>
  </styles__Container>
</Memo(ImageWithPlaceholder)>
`;

exports[`<ImageWithPlaceholder /> should render 2`] = `
<Memo(ImageWithPlaceholder)
  imageUrl="https://commons.wikimedia.org/wiki/File:Old_version_of_the_Wikipedia_logo_used_until_2010_(big,_English).png"
>
  <styles__Container>
    <div
      className="styles__Container-sc-1li6v9v-0 HMjkw"
    >
      <ReactPlaceholder
        firstLaunchOnly={true}
        ready={false}
        showLoadingAnimation={true}
        style={
          Object {
            "marginRight": "0",
            "position": "absolute",
          }
        }
        type="rect"
      >
        <RectShape
          className="show-loading-animation"
          color="#CDCDCD"
          rows={3}
          style={
            Object {
              "marginRight": "0",
              "position": "absolute",
            }
          }
        >
          <div
            className="rect-shape show-loading-animation"
            style={
              Object {
                "backgroundColor": "#CDCDCD",
                "height": "100%",
                "marginRight": "0",
                "position": "absolute",
                "width": "100%",
              }
            }
          />
        </RectShape>
      </ReactPlaceholder>
      <Image
        layout="fill"
        objectFit="cover"
        onLoad={[Function]}
        quality={75}
        src="https://commons.wikimedia.org/wiki/File:Old_version_of_the_Wikipedia_logo_used_until_2010_(big,_English).png"
      >
        <div
          style={
            Object {
              "bottom": 0,
              "boxSizing": "border-box",
              "display": "block",
              "left": 0,
              "margin": 0,
              "overflow": "hidden",
              "position": "absolute",
              "right": 0,
              "top": 0,
            }
          }
        >
          <img
            decoding="async"
            onLoad={[Function]}
            src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
            style={
              Object {
                "border": "none",
                "bottom": 0,
                "boxSizing": "border-box",
                "display": "block",
                "height": 0,
                "left": 0,
                "margin": "auto",
                "maxHeight": "100%",
                "maxWidth": "100%",
                "minHeight": "100%",
                "minWidth": "100%",
                "objectFit": "cover",
                "objectPosition": undefined,
                "padding": 0,
                "position": "absolute",
                "right": 0,
                "top": 0,
                "visibility": "hidden",
                "width": 0,
              }
            }
          />
        </div>
      </Image>
    </div>
  </styles__Container>
</Memo(ImageWithPlaceholder)>
`;
